Feature: broader uses for watch plugins

tdd opened the pull request:
## Summary

As discussed in #6467, watch plugins could be much more useful if they could change a wider range of global config options. This implements a white-listed wider set, tests and docs it.

Two immediate use cases are toggler plugins for tests verbosity and coverage collection. They're already ready and tested on my local machine, but I have no idea how to get them inside the `jest-community` org, and I'll need the relevant minimum Jest version for a pertinent `peerDependency`.

## Test plan

Currently `master` fails to build on its own, but from what I could see I do not introduce any extra failure. Extra tests are located in `packages/jest-cli/__tests__/watch.test.js`, and focus on confirming white-listing of allowed global config options is effective.

The docs in `/docs` were also updated (but they don't seem to be used by `website/`'s `yarn start` or `yarn build`, so I only checked their rendering in VS Code's Markdown preview).

This is not semver-major, as there are no breaking changes IMHO, this is more of an extra feature, so looks semver-minor to me.

Hope you like it!

Closes #6467

tdd commented:
@rogeliog this is for your reviewing pleasure 😉

SimenB commented on docs/WatchPlugins.md:
```diff
@@ -148,3 +148,24 @@ class MyWatchPlugin {
+
+#### Authorized configuration keys
+
+For stability and safety reasons, only part of the global configuration keys can be updated with `updateConfigAndRun`. The current white list is as follows:
```
https://facebook.github.io/jest/docs/en/cli.html#changedsince, https://facebook.github.io/jest/docs/en/cli.html#lastcommit and https://facebook.github.io/jest/docs/en/cli.html#onlychanged would be awesome as well. Maybe https://facebook.github.io/jest/docs/en/cli.html#logheapusage?

tdd replied:
Hey @SimenB!

- `--changedSince` and `--lastCommit` do not have any direct equivalent in the global configuration object. They're used by `getChangedFilesForRoots` in `jest-changed-files`, and are also ignored when `--onlyChanged` (which is the default watch mode), so they don't seem to make sense in that mode.  At any rate, the watch architecture would not let us change that on the fly right now, this is a bootstrap-time computation.
- `--onlyChanged` is already togglable in watch mode: these are the `a` key (all tests, same as `--watchAll`) vs. `o` key (only changed tests, same as `--onlyChanged`).
- `--logHeapUsage` does have a matching boolean in the global config, so I'm open to adding it, what do others think?

SimenB replied:
in the same way toggling `onlyChanged` is useful, I think `changedSince` would be useful. It would then overwrite `onlyChanged`.

tdd replied:
@SimenB I agree it would be nice, but Jest's current architecture doesn't let us do it. It computes the watched list at boot time, then keeps it in a closure where specific global config options may amend it (such as the watch mode). There's no stuff in there right now for `changedSince` or `lastCommit`, and I feel the change would be significant.

This specific PR is bound by what's in the global config. Should this feature set become config-based, I'll be glad to expand the white list, but it's a different feature scope IMHO.

SimenB replied:
@rogeliog sounds like we need another of your magical refactor runs 😀

rogeliog replied:
Yeah we would need to change a couple of things in order to make that happen 🙃

SimenB commented on packages/jest-cli/src/__tests__/watch.test.js:
```diff
@@ -391,6 +396,119 @@ describe('Watch mode flows', () => {
+      {virtual: true},
+    );
+
+    // updateGlobalConfig.mockReset();
```
?

tdd replied:
Woops, forgot to strip this 😒

SimenB commented on packages/jest-cli/src/__tests__/watch.test.js:
```diff
@@ -86,6 +86,11 @@ jest.doMock(
   {virtual: true},
 );
 
+const regularUpdateGlobalConfig = require('../lib/update_global_config')
```
You can `import` it to avoid the `.default` call

tdd replied:
Yes, but as I'm mocking it right the next line, it seemed more interesting to keep it close by?

thymikee replied:
`jest.doMock` should be called prior to require call

SimenB commented:
@tdd did you see the CLA comment?

tdd commented:
@SimenB I did see the CLA comment and e-mailed to update and renew my CLA status. The online form fails in obscure ways for updates (always has, nobody's fixing it).

rogeliog commented on packages/jest-cli/src/__tests__/watch.test.js:
```diff
@@ -391,6 +396,119 @@ describe('Watch mode flows', () => {
+    );
+
+    // updateGlobalConfig.mockReset();
+    const whiteListedOptions = [
```
I think we could use `it.each([objectWithWhitelistedOptionsAndOtherOptions])('')`

https://facebook.github.io/jest/docs/en/api.html#testeachtable-name-fn

tdd replied:
Right! Good point…  Let me try…

rogeliog commented on packages/jest-cli/src/watch.js:
```diff
@@ -68,19 +72,71 @@ export default function watch(
+    // It is unfortunate that Flow isn't smart enough yet to let us
+    // implement a generic assign loop over a list of white-listed
+    // option names here :-(
     globalConfig = updateGlobalConfig(globalConfig, {
```
Now that there are a lot of options, maybe we could move a lot of this `!== undefined` logic into the `updateGlobalConfig` function, so that it is the only place that knows how to handle all of that logic

tdd replied:
Yeah, that's an option. I really disliked the un-DRYness of this part of the code, but making it dynamically iterate over a list of props confused the heck out of Flow 😒 .

I'll try and move that logic inside `updateGlobalConfig` and see how it goes.

tdd commented:
Hey @rogeliog!

Here's my update. It addresses the various suggestions you had, and it looks like it's going to greenlight through. Let me know what the next steps would be!

tdd commented:
Hey @rogeliog just following up on this. Have you had time to peruse the latest changes? I'd love to know how and when we can move ahead with this 😉

tdd commented:
Just pushed a force update to fix upstream merge conflicts over `CHANGELOG.md` after having rebased on top of it.

thymikee commented:
@tdd mind resolving conflits? :)

tdd commented:
@thymikee dang, there were none when this was originally approved. Also, all checks passed. Checking it now.

tdd commented:
Grmbl, this thing stayed too long in limbo, approved but unmerged, as other PRs for watch plugins came in that broke this one. Will need to investigate. (Also, not cool, team 😐 )

thymikee commented:
I'm pretty sure @rogeliog can help you with figuring out what's wrong now :). 
Thanks for your patience though! We try our best to review and merge stuff as early as possible, but there's a lot to do 😅.

tdd commented:
Hey @thymikee I sure would appreciate @rogeliog's help on this, the issue is the same across the board, when `require`'ing watch plugins' paths. `jest-resolve` b0rks further down the line. This is not code I touched in my PR and I have no idea just now about why the path would be missing here—and not break `watch.js` in `master`?

rogeliog commented:
Yes, I'm more than happy to help. I'll take a look

tdd commented:
@rogeliog thank you that's great! Any idea when this could get merged in, @thymikee?

Also, does any of you know how I could put two upcoming plugins (that need this to come in), one for [toggling verbosity](https://github.com/deliciousinsights/jest-watch-toggle-verbosity) and one for [toggling coverage](https://github.com/deliciousinsights/jest-watch-toggle-coverage), into jest-community? They're in my own account for now (along with [chai-jest-diff](https://github.com/deliciousinsights/chai-jest-diff), for instance), but they'd be better out there.

thymikee commented:
Hopefully sometime soon! I'll try to review this today.

You'll need to transfer ownership of the repositories to jest-community. Then we'll be able to make you a collaborator on these projects, so you still can be in charge of them. cc @SimenB

tdd commented:
Alright, the moment a public release ships with this merged in, I'll
update the plugins, test again, then transfer.

On Sun, Jul 8, 2018, at 16:01, Michał Pierzchała wrote:
> Hopefully sometime soon! I'll try to review this today.


> You'll need to transfer ownership of the repositories to jest-
> community. Then we'll be able to make you a collaborator on these
> projects, so you still can be in charge of them. cc @SimenB[1]> — You are receiving this because you were mentioned. Reply to this
> email directly, view it on GitHub[2], or mute the thread[3].
--
Christophe Porteneuve
tdd@tddsworld.com

Links:

  1. https://github.com/SimenB
  2. https://github.com/facebook/jest/pull/6473#issuecomment-403289802
  3. https://github.com/notifications/unsubscribe-auth/AAAOqMXpBnCKuoHRsb-zzkJ3LAU6kaHqks5uEhCtgaJpZM4UpZAi

thymikee commented on packages/jest-cli/src/lib/update_global_config.js:
```diff
@@ -7,16 +7,32 @@
+} from 'types/Config';
 
 type Options = {
+  bail?: boolean,
```
Options is a subset of GlobalConfig. So we could use it as source of truth for our types by leveraging [`$PropertyType<T, k>`](https://flow.org/en/docs/types/utilities/#toc-propertytype) flow operator:
```jsx
type Options = {
  bail?: $PropertyType<GlobalConfig, 'bail'>,
  ...
}
```
(not mandatory for this PR though, just though you'd like to know about it)

tdd replied:
@thymikee Interesting! Not knowledgeable enough about Flow so far to have seen this one coming. I'm updating the `Options` type to ref `GlobalConfig` types whenever applicable (anything but `mode`, basically)

thymikee commented on packages/jest-cli/src/watch.js:
```diff
@@ -68,31 +71,55 @@ export default function watch(
     testNamePattern?: string,
     testPathPattern?: string,
     updateSnapshot?: SnapshotUpdateState,
+    verbose?: boolean,
```
Looks like this could be replaced with `Options` type from `update_global_config.js`?

tdd replied:
@thymikee By now this is correct — the patch evolved organically, but now we're done! Centralizing.

rogeliog commented:
@tdd I love the idea of these plugins... I have some idea of what we could do to prevent having a separate plugin for each config option. As of `jest@23.3.0` watch plugins support configs(https://jestjs.io/docs/en/watch-plugins#customization). It would be awesome to just have one plugin `jest-watch-toggle-config`. And can be used like this.

```json
"watchPlugins": [
  ["jest-watch-toggle-config", { "toggle": "verbose", "key": "v" }],
  ["jest-watch-toggle-config", { "toggle": "coverage" }],
  ["jest-watch-toggle-config", { "toggle": "notify", "key": "n" }]
]
```

we could add a default `key` and `prompt` for each toggle config option, while still allowing the user to override those.... thoughts?

SimenB commented:
Or even something like https://www.npmjs.com/package/prompt-checkbox, where `checked` means `true`, and you could toggle all in the same view

tdd commented:
I think making it generic is a nice approach, but I can imagine
scenarios where the toggling is not just true/false, but null/undefined
vs. Something more involved, like an array or string. So the prompt
approach wouldn't be great for that. Also, while I appreciate reserving
a single key, I think a lot of users want single-keypress behavior, not
just opening up a UI, especially for toggling. So I guess multiple
instances with their own keys and parameters values would be better.
I'll try to whip out an implementation later this week, if my
newborn lets me 😁

On Mon, Jul 9, 2018, at 07:45, Simen Bekkhus wrote:
> Or even something like https://www.npmjs.com/package/prompt-checkbox,
> where checked means true, and you could toggle all in the same view> — You are receiving this because you were mentioned. Reply to this
> email directly, view it on GitHub[1], or mute the thread[2].
--
Christophe Porteneuve
tdd@tddsworld.com

Links:

  1. https://github.com/facebook/jest/pull/6473#issuecomment-403366235
  2. https://github.com/notifications/unsubscribe-auth/AAAOqE90Flm-gA7mg-RKAbfPthno3rRrks5uEu4KgaJpZM4UpZAi

tdd commented:
Hey @rogeliog @thymikee here's my take at generic boolean setting toggling through watch plugin config: https://github.com/deliciousinsights/jest-watch-toggle#readme

I'll strip the more-specific repos.

tdd commented:
Now that I think of it: there's another reason I'm not a fan of @SimenB's suggestion of using an interactive checkbox list UI for toggling settings: this means the list of settings is hardcoded in the plugin. So when Jest whitelists more, the plugin needs maintenance. I'd rather let the end user do the config for them to (a) gain single-key toggling and (b) leverage any latest whitelisted setting.

tdd commented:
@rogeliog I like how your demo config for the single plugin lets end users write terse configs by ensuring proper default prompts and keys, but I'm concerned that:

- this means hard-coding, and therefore maintaining over time, a list of whitelisted boolean update-config options (we could throw on plugin-unknown settings with no provided key or prompt, but still)
- this may conflict with later built-in keys in Jest watch mode (there are more and more…)

Also, I missed your initial name proposal and pushed a `jest-watch-toggle` repo for now. Do you think `…toggle-config` is better? Haven't pushed to npm yet anyway, as the matching Jest isn't there yet.

tdd commented:
Dammit, something's wrong with the `jest-website` command apparently.

tdd commented:
Just to be clear: master isn't always passing in Jest, is it?

thymikee commented:
Website failures are unrelated, it's probably some temporary outage on OpenCollective API.

tdd commented:
Well, 23.4 shipped without this still… Hence CHANGELOG conflicts. Addressing this.

thymikee commented:
@SimenB can you have a final look and merge if all good?

thymikee commented:
I recall @rogeliog having some thoughts around that. To me it's perfectly fine and seems like good DX.

tdd commented:
@thymikee now that I think of it, we shouldn't put that burden on plugin authors, when Jest itself can check and bail.

SimenB commented:
We explicitly want to allow overrides of built-ins, at least. For e.g. https://github.com/jest-community/jest-watch-typeahead

tdd commented:
@SmienB that's a good point. Perhaps the new code in `watch.js` would differentiate like this:

- keys reserved by built-in plugins that provide core functionality (think `q` for `quit.js` 😉): not overridable
- keys reserved by built-in plugins that provide extra functionality (e.g. `test_name_pattern.js`): overridable
- keys reserved by prior third-party plugins: not overridable

We *could* provide a config option that lets us enable full-scale override, but that seems like providing users with a gun whose only target is their own foot.

What do you think?

rogeliog commented:
I agree with a lot of the comments above, I think we need to find a good way to handle this.

Something within Jest that knows how to handle this conflicts and give the appropriate error message if needed. For plugin authors and consumers, things should just work.

tdd commented:
@rogeliog about the toggle-config plugin: I decided to go with…

1. `jest-watch-toggle-config` package name
2. Sane defaults for the four boolean config options that seemed to make sense in watch mode (and not be redundant with other builtin keys): `bail`, `collectCoverage`, `notify` and `verbose`

[See the plugin here](https://github.com/deliciousinsights/jest-watch-toggle-config#readme) (assumes 23.5 will have this PR merged in, not published to npm or moved to jest-community until this lands in a released version)

As for Jest's better DX wrt conflicting keys, I'll set up an issue soon with a detailed spec proposal and ping you all on it; if the spec gels, I'll put up a PR for it.

tdd commented:
🎉🙏


On Thu, Jul 12, 2018, at 08:27, Simen Bekkhus wrote:
> Merged #6473[1].


> — You are receiving this because you were mentioned. Reply to this
> email directly, view it on GitHub[2], or mute the thread[3].
--
Christophe Porteneuve
tdd@tddsworld.com

Links:

  1. https://github.com/facebook/jest/pull/6473
  2. https://github.com/facebook/jest/pull/6473#event-1729518014
  3. https://github.com/notifications/unsubscribe-auth/AAAOqMesNg6Us9D-sCPYaJ9Stppg0uuzks5uFuxGgaJpZM4UpZAi

tdd commented:
Hey all! FWIW the initial release of the toggle-config plugin is out, now that 23.4.1 includes this.

Will move it to jest-community soon. Check it out: https://github.com/deliciousinsights/jest-watch-toggle-config#readme

tdd commented:
Hey @rogeliog @SimenB I added you as admins on my toggle-config plugin so you would transfer it to `jest-community`. Once you've done so and added me back as contributor to it, I'll update its `package.json`, set up its Travis CI badge, semver-patch it and republish.

SimenB commented:
Done, and added you as admin

tdd commented:
Awesome, thank you!
